[cometvisu] Auto-install CometVisu-Client feature added + some minor changes

peuter opened the pull request:
- change generator destination folder to fit the new structure
- support the new Location type
- introduce ClientInstaller for checking for installed CometVisu client and auto-install the latest release if there is no client

Signed-off-by: Tobias Bräutigam <tbraeutigam@gmail.com>

<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

<!-- TITLE -->

<!--
Please provide a PR summary in the *Title* above, according to the following schema:
- If related to one specific addon: Mention the addon shortname in square brackets
  e.g. "[exec]", "[netatmo]" or "[tesla]"
- If the PR is work in progress: Add "[WIP]"
- Give a short meaningful description in imperative mood
  e.g. "Add support for device XYZ" or "Fix wrongly handled exception"
  for a new add-on/binding: "Initial contribution"
Examples:
- "[homematic] Improve communication with weak signal devices"
- "[timemachine][WIP] Initial contribution"
- "Update contribution guidelines on new signing rules"
-->

<!-- DESCRIPTION -->

<!--
Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the addon README?
- Does your contribution follow the coding guidelines:
  http://docs.openhab.org/developers/development/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  http://docs.openhab.org/developers/development/bindings.html#static-code-analysis
- Did you sign-off your work:
  http://docs.openhab.org/developers/contributing/contributing#sign-your-work
-->

<!-- TESTING -->

<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It's a good practice to add an URL to your built JAR in this Pull Request description,
so it's easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it's reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

kaikreuzer commented:
How does this feature look like for the user?
Note that openHAB must not download stuff from the net without the explicit approval by the user.
I would therefore expect at least a configuration parameter that allows to switch this feature on and off. At the same time, the user should be informed about the license of the downloaded software.

peuter commented:
I wasn't aware of the download restriction. I just thought, that the user has to download the CometVisu client anyways to make it work, I could ease things here.

I added a config option now (which is disabled by default) which can enable the automatic download.
I also changed the 404 page, which is shown when the user has not yet downloaded the CometVisu client.
Its shows some configuration settings and a button to execute the download (and store the changed settings). If the user clicks on that button a popup with licensing information shows up. I hope this is enough.

B.t.w I am looking for a well thought default value for the ``webFolder`` configuration setting (Note: this is the folder where the CometVisu client is downloaded to when it does not exist). It must be a folder openHAB has write access to. I am not familiar with the openHAB runtimes folder structure. Is there a path (relative to the runtimes root?) i could use as default value?

Please do not merge yet, just found out that something is missing. I will add test bundle to this PR once I am done.

peuter commented:
Ok, now I'm done. Whats left to to is to change the default setting for ``webFolder``. Currently its ``/var/www/cometvisu`` which will not work in every environment (e.g. Windows machines or no access rights to that folder.

@kaikreuzer If you have a better suggestion for a default value that should work on every openHAB installation, please let me now.

Here is the bundle for testing purposes.
[org.openhab.ui.cometvisu-2.2.0-SNAPSHOT.zip](https://github.com/openhab/openhab2-addons/files/1219777/org.openhab.ui.cometvisu-2.2.0-SNAPSHOT.zip)

martinvw commented on addons/ui/org.openhab.ui.cometvisu/about_files/ANOTHER-REST-CLIENT_LICENSE:
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
Can you refer to a remote location instead, I don't think is available when packaged.

peuter replied:
I just did it like it's done in `org.eclipse.smarthome.ui.paper`, so I thought this would be the preferred way. No problem will change that.

martinvw replied:
I could also be wrong @kaikreuzer...?

kaikreuzer replied:
Either about_files should be added to build.properties so that it becomes available in the binary or an external link should be used.
Looking at the MIT license, a link to https://opensource.org/licenses/MIT probably won't do, because the year and copyright holder is expected in the title, so the link should point to the license of the project itself. In case there is a risk that the external link breaks (because the project might not be well structured and often changing), including the license might be the safer option.

martinvw replied:
Thanks!

martinvw commented on addons/ui/org.openhab.ui.cometvisu/src/main/java/org/openhab/ui/cometvisu/backend/ConfigResource.java:
```diff
@@ -0,0 +1,49 @@
+    @Produces({ MediaType.TEXT_PLAIN })
+    public Response getValue(@PathParam("actionName") String actionName) {
+
+        if (actionName.equalsIgnoreCase("download-client")) {
```
Position literals first in String comparisons - that way if the String is null you won't get a NullPointerException, it'll just return false. [source: PMD](http://pmd.sourceforge.net/pmd-4.3.0/rules/design.html#PositionLiteralsFirstInComparisons)

martinvw commented on addons/ui/org.openhab.ui.cometvisu/src/main/java/org/openhab/ui/cometvisu/internal/Config.java:
```diff
@@ -37,13 +38,16 @@
+    public static final String COMETVISU_AUTODOWNLOAD_PROPERTY = "autoDownload";
 
-    public static String COMETVISU_WEBFOLDER = "runtime/web/";
+    public static String COMETVISU_WEBFOLDER = "/var/www/cometvisu";
```
I think Kai did not yet respond on the location but we had such a question before, it should be located under *userdata* and there should be a method to retrieve the location.

@kaikreuzer correct me if I'm wrong and if you know the method for the location please mention 😄

kaikreuzer replied:
A hard coded absolute address definitely looks baaaad :-)
If this is the folder where the runtime automatically downloads stuff to, I would agree it should be the userdata folder (as this is the only one officially writable). To get the location from code `ConfigConstants.getUserDataFolder()` should be used.

martinvw commented on addons/ui/org.openhab.ui.cometvisu/src/main/java/org/openhab/ui/cometvisu/internal/util/ClientInstaller.java:
```diff
@@ -0,0 +1,306 @@
+ */
+public class ClientInstaller {
+
+    static final Logger logger = LoggerFactory.getLogger(ClientInstaller.class);
```
Please do not declare `static` but declare it `private`

martinvw commented on addons/ui/org.openhab.ui.cometvisu/src/main/java/org/openhab/ui/cometvisu/internal/util/ClientInstaller.java:
```diff
@@ -0,0 +1,306 @@
+
+    private static final byte[] buffer = new byte[0xFFFF];
+
+    private Map latestRelease = null;
```
Please use the proper generics and do not initialize with null, it will be done automatically.

martinvw commented on addons/ui/org.openhab.ui.cometvisu/src/main/java/org/openhab/ui/cometvisu/internal/util/ClientInstaller.java:
```diff
@@ -0,0 +1,306 @@
+    /** Regular expression to parse semver version strings */
+    private final Pattern semverPattern = Pattern.compile("[\\D]*([\\d]{1,3})\\.([\\d]{1,3})\\.([\\d]{1,3}).*");
+
+    private ArrayList<String> alreadyCheckedFolders = new ArrayList<String>();
```
1. Please use the diamond operator, ie replace the second occurrence of the same generic by `<>`
2. Declare the type as `List<String>`

martinvw commented on addons/ui/org.openhab.ui.cometvisu/src/main/java/org/openhab/ui/cometvisu/internal/util/ClientInstaller.java:
```diff
@@ -0,0 +1,306 @@
+     * @param force {boolean} force downloading if client not found
+     */
+    public void check(boolean force) {
+        if (alreadyCheckedFolders.contains(Config.COMETVISU_WEBFOLDER) && force == false) {
```
I would prefer `!force`

martinvw commented on addons/ui/org.openhab.ui.cometvisu/src/main/java/org/openhab/ui/cometvisu/internal/util/ClientInstaller.java:
```diff
@@ -0,0 +1,306 @@
+     *
+     * @return {Map}
+     */
+    private Map getLatestRelease() {
```
Please return a generic type

martinvw replied:
How why is the latest release a Map?

peuter replied:
The map is the parsed JSON response from the github API: https://developer.github.com/v3/repos/releases/#get-the-latest-release

martinvw commented on addons/ui/org.openhab.ui.cometvisu/src/main/java/org/openhab/ui/cometvisu/internal/util/ClientInstaller.java:
```diff
@@ -0,0 +1,306 @@
+                    InputStream is = null;
+                    OutputStream os = null;
+
+                    try {
```
Consider using try-with-resources this could simplify the closing of the connections

peuter commented:
I did all the requested changes, despite the license link stuff. Waiting for @kaikreuzer final response.
But I have no idea why the automatic code-review is complaining about a commit with a missing Signed-off-by.

kaikreuzer commented on addons/ui/org.openhab.ui.cometvisu/README.md:
```diff
@@ -11,16 +11,26 @@ This adds a backend for the web based visualization CometVisu (http://www.cometv
 ### Server configuration
 
-You can customize some setting of the CometVisu backend in a configuration file named 'cometvisu.cfg' which has to be created in the 'conf/services/' folder of your openHAB2 instance, if it does not exist.
+#### Version >= 2.2.0
```
The docs should always only describe the latest version - after all, that's why they are in a versioning system.

peuter commented:
Ok, all requested changes are done.

kaikreuzer commented on addons/ui/org.openhab.ui.cometvisu/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,28 @@
+		<parameter name="webFolder" type="text">
+            <label>Filesystem path</label>
+            <description>Defines the absolute path to the CometVisu client directory.</description>
+            <default>userdata/cometvisu</default>
```
I would remove the default here. Firstly, it is NOT an absolute path (although the description says it has to be) and secondly, the path is a different one on apt installations.

kaikreuzer commented on addons/ui/org.openhab.ui.cometvisu/ESH-INF/config/config.xml:
```diff
@@ -0,0 +1,28 @@
+
+	<config-description uri="ui:cometvisu">
+		<parameter name="webFolder" type="text">
+            <label>Filesystem path</label>
```
Please re-apply the formatter, here is a tabs/spaces mix.

kaikreuzer commented on addons/ui/org.openhab.ui.cometvisu/README.md:
```diff
@@ -11,16 +11,22 @@ This adds a backend for the web based visualization CometVisu (http://www.cometv
+
+### Via Config file
+
+You can customize some settings of the CometVisu backend in a configuration file named 'cometvisu.cfg' which has to be created in the 'conf/services/' folder of your openHAB2 instance, if it does not exist.
```
Change "openHAB2" to "openHAB".

kaikreuzer commented on addons/ui/org.openhab.ui.cometvisu/src/main/java/org/openhab/ui/cometvisu/servlet/CometVisuApp.java:
```diff
@@ -57,12 +58,20 @@
     private EventPublisher eventPublisher;
 
-    private CometVisuServlet servlet;
+    public static CometVisuServlet servlet;
```
Why does it have to be static? Sould be avoided, if possible.

kaikreuzer commented on addons/ui/org.openhab.ui.cometvisu/src/main/java/org/openhab/ui/cometvisu/servlet/CometVisuApp.java:
```diff
@@ -57,12 +58,20 @@
 
     protected static Map<String, QueryablePersistenceService> persistenceServices = new HashMap<>();
 
+    public final static ClientInstaller installer = new ClientInstaller();
```
Why does it have to be static? Sould be avoided, if possible.

kaikreuzer commented on addons/ui/org.openhab.ui.cometvisu/src/main/java/org/openhab/ui/cometvisu/servlet/CometVisuApp.java:
```diff
@@ -57,12 +58,20 @@
 
+    public final static ClientInstaller installer = new ClientInstaller();
+
+    private static Map<String, Object> properties = new HashMap<>();
```
Why does it have to be static? Sould be avoided, if possible.

kaikreuzer commented on addons/ui/org.openhab.ui.cometvisu/src/main/java/org/openhab/ui/cometvisu/servlet/CometVisuApp.java:
```diff
@@ -57,12 +58,20 @@
+
+    private static Map<String, Object> properties = new HashMap<>();
+
+    public static Map<String, Object> getProperties() {
```
Why does it have to be static? Sould be avoided, if possible.

peuter commented:
And again, all requested changes done ;-)
